ecl/hql/hqlgram2.cpp

```diff
@@ -12329,30 +12335,34 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
         throw;
     }
     OwnedHqlExpr parsed = scope->lookupSymbol(name, LSFsharedOK|LSFnoreport, ctx);
-    bool canCache = false;
-    bool isMacro = false;
-    bool neverSimplify = false;
-    if (parsed)
-    {
-        canCache = (parsed->getOperator() != no_forwardscope) && !alreadySimplified ;
-        isMacro = parsed->isMacro();
-        neverSimplify = isMacro || ctx.neverSimplify(fullName);
-    }
     ctx.incrementAttribsProcessed();
-    if (canCache && (ctx.syntaxChecking() || ctx.hasCacheLocation()))
+
+    if (parsed && !alreadySimplified)
     {
-        OwnedHqlExpr simplified = neverSimplify ? nullptr : createSimplifiedDefinition(parsed);
-        if (ctx.hasCacheLocation())
-            attrCtx.createCache(simplified, isMacro);
-        if (simplified)
+        const bool canCache = parsed->getOperator() != no_forwardscope;
+        if (canCache && (ctx.hasCacheLocation() || ctx.syntaxChecking()))
         {
-            ctx.incrementAttribsSimplified();
+            const bool isMacro = parsed->isMacro();
+            OwnedHqlExpr simplified;
+            if (!isMacro && (!ctx.ignoreSimplified() || ctx.checkSimpleDef()) && !ctx.neverSimplify(fullName))
```

Comment:
I'm still not convinced this matches the code below.  I would imagine

if (!isMacro &&
   ((ctx.hasCacheLocation() && !ctx.ignoreSimplified) ||
    ctx.checkSimpleDef() ||
    (ctx.syntaxChecking() && !ctx.ignoreSimplified()) && ctx.neverSimplify(fullName))

You could also argue that !ignoreSimplified() is also checked for the middle test.  If so it would become:

if (!isMacro && !ctx.ignoreSimplified() && 
   ((ctx.hasCacheLocation() ||
    ctx.checkSimpleDef() ||
    ctx.syntaxChecking())  && ctx.neverSimplify(fullName))

I need to think about whether there is an issue with the cache already being up to date, but I think that is only relevant to your other PR.

Reply:
The second test looks neat and easily comprehensible.  But just to be clear we did previously agree that "If ctx.ignoreSimplified() is true and checkSimpleDef is true, then it needs to generate a simplified expression" so that it can check the simplified expression.  As long as we happy that these 2 flags set true means that the simplified expression will not be verified, we can use the second test.